// Copyright (c) 2015 Urtashev Arslan. All rights reserved.
// Contacts: <urtashev@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software
// and associated documentation files (the "Software"), to deal in the Software without
// restriction, including without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
//   The above copyright notice and this permission notice shall be included in all copies or
//   substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

syntax = "proto2";
option cc_enable_arenas = true;

package geo_base.proto;

message location_t {
    required double lat = 1;
    required double lon = 2;
}

message polygon_t {
    required uint64 polygon_id = 1;
    repeated location_t locations = 2;

    enum type_t {
        TYPE_UNKNOWN = 0;
        TYPE_INNER = 1;
        TYPE_OUTER = 2;
    }

    required type_t type = 3;
}

message kv_t {
    required string k = 1;
    required string v = 2;
}

message region_t {
    required uint64 region_id = 1;
    optional uint64 parent_id = 2;
    repeated polygon_t polygons = 3;
    repeated kv_t kvs = 4;
    repeated string blobs = 5;

    enum option_t {
        OPTION_ZERO = 0;
        OPTION_BOUNDARY_ADMINISTRATIVE = 1;
        OPTION_PLACE_ISLAND = 2;
        OPTION_PLACE_TOWN = 3;
        OPTION_PLACE_CITY = 4;
        OPTION_PLACE_VILLAGE = 5;
        OPTION_PLACE_BOROUGH = 6;
        OPTION_PLACE_SUBURB = 7;
        OPTION_PLACE_CONTINENT = 8;
    }

    required uint64 options = 6;
}
